### PR TITLE
Improve list projection to take a semtype as the key

### LINKED ISSUE #891: Support for typing list projection with int subtype
We need to be able to compute the type of `v[i]` when v is a subtype of list and i is an arbitrary subtype of int (not just `int` or a `singleton`).

Needed for #576.

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
Fixes https://github.com/ballerina-platform/nballerina/issues/891

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/types/core.bal [hunk 1]
```diff
@@ -945,13 +945,17 @@ function bddListAtomicType(Env env, Bdd bdd, ListAtomicType top) returns ListAto
 // This is what Castagna calls projection.
 // We will extend this to allow `key` to be a SemType, which will turn into an IntSubtype.
 // If `t` is not a list, NEVER is returned
-public function listMemberType(Context cx, SemType t, int? key = ()) returns SemType {
+public function listMemberType(Context cx, SemType t, SemType? k = INT) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/core.bal]: `SemType` not `SemType?`

File: compiler/modules/types/core.bal [hunk 2]
```diff
@@ -945,13 +945,17 @@ function bddListAtomicType(Env env, Bdd bdd, ListAtomicType top) returns ListAto
...
     if t is UniformTypeBitSet {
         return (t & LIST) != 0 ? TOP : NEVER;
     }
     else {
-        return union(bddListMemberType(cx, <Bdd>getComplexSubtypeData(t, UT_LIST_RO), key, TOP),
-                     bddListMemberType(cx, <Bdd>getComplexSubtypeData(t, UT_LIST_RW), key, TOP));
+        IntSubtype? intSubtype = ();
+        if k is ComplexSemType {
+            intSubtype = <IntSubtype>getComplexSubtypeData(k, UT_INT);
```
Comment (jclark) [inline compiler/modules/types/core.bal]: We should use new `intSubtype` function here (and for mappingMemberType).
It's a bit awkard that usually we mean `true` for all ints, but here we are using `()` for all ints. Can we do better?

File: compiler/modules/types/list.bal [hunk 3]
```diff
@@ -217,9 +217,16 @@ function listIntersectWith(FixedLengthArray members, SemType rest, ListAtomicTyp
...
-            intersect(listMemberAt(members, rest, i), listMemberAt(lt.members, lt.rest, i)));
-    } 
+        fixedArraySet(members, i, intersect(
+                                        listMemberAt(members, rest, i),
+                                        listMemberAt(lt.members, lt.rest, i)));
+    }
+    // Intersect the last member of initial array
+    if nonRepeatedLen < members.fixedLength {
+        members.initial[nonRepeatedLen] =  intersect(
+                                                listMemberAt(members, rest, nonRepeatedLen),
```
Comment (jclark) [inline compiler/modules/types/list.bal]: I typically don't format calls like this: I put the start of the first arg on the same line as the `(`

File: compiler/modules/types/list.bal [hunk 4]
```diff
@@ -255,7 +262,7 @@ function listInhabited(Context cx, FixedLengthArray members, SemType rest, ListC
             if listInhabited(cx, members, NEVER, neg.next) {
                 return true;
             }
-            foreach int i in len + 1 ..< negLen {
+            foreach int i in members.initial.length() + 1 ..< nt.members.initial.length() {
```
Comment (jclark) [inline compiler/modules/types/list.bal]: Why are we changing this?

File: compiler/modules/types/list.bal [hunk 5]
```diff
@@ -334,7 +341,16 @@ function fixedArrayFill(FixedLengthArray arr, int newLen, SemType filler) {
     if newLen <= initial.length() {
         return;
     }
-    if arr.fixedLength == 0 || initial[initial.length() - 1] != filler {
+    int initLen = initial.length();
```
Comment (jclark) [inline compiler/modules/types/list.bal:354]: This was buggy before, right? Do we have a test case?

File: compiler/modules/types/int.bal [hunk 6]
```diff
@@ -133,6 +133,21 @@ function intSubtypeComplement(SubtypeData d) returns SubtypeData {
     return rangeListComplement(v).cloneReadOnly();
 }
 
+// `min` and `max` are inclusive
+function intSubtypeOverlapRange(IntSubtype subtype, int min, int max) returns boolean {
+    IntSubtype range = [{ min, max }];
+    return intSubtypeIntersect(subtype, range) != false;
+}
+
+function intSubtypeContainsGreaterThan(IntSubtype subtype, int value) returns boolean {
```
Comment (jclark) [inline compiler/modules/types/int.bal]: Can we instead have:

```
function intSubtypeMax(IntSubtype subtype) returns int {
    return subtype[subtype.length() - 1].max;
}
```

Similarly for intSubtypeMin. Then use in intConstraints function in core.

File: compiler/modules/types/int.bal [hunk 7]
```diff
@@ -133,6 +133,21 @@ function intSubtypeComplement(SubtypeData d) returns SubtypeData {
     return rangeListComplement(v).cloneReadOnly();
 }
 
+// `min` and `max` are inclusive
+function intSubtypeOverlapRange(IntSubtype subtype, int min, int max) returns boolean {
```
Comment (jclark) [inline compiler/modules/types/int.bal]: Slightly prefer to pass in a Range argument.

File: compiler/modules/types/listProj.bal [hunk 8]
```diff
@@ -67,39 +68,44 @@ function listProjPath(Context cx, int k, Conjunction? pos, Conjunction? neg) ret
...
         int len = members.fixedLength;
-        ListAtomicType nt = cx.listAtomType(neg.atom);
+        ListAtomicType nt = neg.listType;
         int negLen = nt.members.fixedLength;
+        SemType p = NEVER;
         if len < negLen {
             if isNever(rest) {
                 return listProjExclude(cx, k, members, rest, neg.next);
             }
+            foreach int i in members.initial.length() + 1 ..< nt.members.initial.length() {
```
Comment (jclark) [inline compiler/modules/types/listProj.bal]: There seems to be a chunk of very similar code in listInhabited, which I don't understand either.

File: compiler/modules/types/listProj.bal [hunk 9]
```diff
@@ -67,39 +68,44 @@ function listProjPath(Context cx, int k, Conjunction? pos, Conjunction? neg) ret
...
         }
         // now we have nt.members.length() <= len
-        SemType p = NEVER;
-        foreach int i in 0 ..< len {
-            SemType ntm = listMemberAt(nt.members, nt.rest, i);
-            SemType d = diff(fixedArrayGet(members, i), ntm);
+        foreach int i in 0 ..< int:max(members.initial.length(), neg.maxInitialLen) {
+            SemType d = diff(listMemberAt(members, rest, i), listMemberAt(nt.members, nt.rest, i));
             if !isEmpty(cx, d) {
                 FixedLengthArray s = fixedArrayShallowCopy(members);
```
Comment (jclark) [inline compiler/modules/types/listProj.bal]: I'm seeing multiple cases of fixedArrayShallowCopy then fill (or fill then set). A helper function might make things clearer.

**Review by jclark — CHANGES_REQUESTED** [review]
I think you are going to need to talk me through the changes to listIhabited.

File: compiler/modules/types/listProj.bal (hunk 9 above)
Comment (rdhananjaya) [reply compiler/modules/types/listProj.bal]: How about `fixedArraySetOnCopy` for the name?

File: compiler/modules/types/list.bal (hunk 4 above)
Comment (rdhananjaya) [reply compiler/modules/types/list.bal]: This part of the code still had the large fixed-length array issue (https://github.com/ballerina-platform/nballerina/issues/820)

File: compiler/modules/types/list.bal (hunk 5 above)
Comment (rdhananjaya) [reply compiler/modules/types/list.bal:354]: I observe this in the debugger, let me add a test

Comment (rdhananjaya) [reply compiler/modules/types/list.bal:354]: Test cases in `proj9-t.bal` are failing without this fix. Specifically `T4[1]`.

File: compiler/modules/types/listProj.bal [hunk 10]
```diff
@@ -1,21 +1,19 @@
...
-        return NEVER;
-    }
+public function listProj(Context cx, SemType t, SemType k) returns SemType {
     if t is UniformTypeBitSet {
         return (t & LIST) != 0 ? TOP : NEVER;
     }
     else {
-        return union(listProjBdd(cx, k, <Bdd>getComplexSubtypeData(t, UT_LIST_RO), (), ()),
-                     listProjBdd(cx, k, <Bdd>getComplexSubtypeData(t, UT_LIST_RW), (), ()));
+        IntSubtype|boolean keyData = intSubtype(k);
```
Comment (jclark) [inline compiler/modules/types/listProj.bal:9]: If it's `false`, then the function will always return `never`, won't it?

So I would pass `IntSubtype|true` around.

**Review by jclark — COMMENTED** [review]
Looks pretty good, but I need some help to understand it all.

File: compiler/modules/types/list.bal [hunk 11]
```diff
@@ -217,9 +217,14 @@ function listIntersectWith(FixedLengthArray members, SemType rest, ListAtomicTyp
...
     int nonRepeatedLen = int:max(members.initial.length(), lt.members.initial.length());
     foreach int i in 0 ..< nonRepeatedLen {
-        fixedArraySet(members, i, 
-            intersect(listMemberAt(members, rest, i), listMemberAt(lt.members, lt.rest, i)));
-    } 
+        fixedArraySet(members, i, intersect(listMemberAt(members, rest, i),
+                                            listMemberAt(lt.members, lt.rest, i)));
+    }
+    // Intersect the last member of initial array
+    if nonRepeatedLen < members.fixedLength {
```
Comment (jclark) [inline compiler/modules/types/list.bal]: I don't understand this.

File: compiler/modules/types/list.bal [hunk 12]
```diff
@@ -255,7 +260,7 @@ function listInhabited(Context cx, FixedLengthArray members, SemType rest, ListC
             if listInhabited(cx, members, NEVER, neg.next) {
                 return true;
             }
-            foreach int i in len + 1 ..< negLen {
+            foreach int i in members.initial.length() + 1 ..< nt.members.initial.length() {
```
Comment (jclark) [inline compiler/modules/types/list.bal]: For upper bound, why not max withneg.initialLen?

For lower bound, why not len?

File: compiler/modules/types/list.bal [hunk 13]
```diff
@@ -366,6 +380,13 @@ function fixedArrayShallowCopy(FixedLengthArray array) returns FixedLengthArray
     return { initial: shallowCopyTypes(array.initial), fixedLength: array.fixedLength };
 }
 
+function fixedArraySetOnCopy(FixedLengthArray array, int index, SemType t, SemType rest) returns FixedLengthArray {
```
Comment (jclark) [inline compiler/modules/types/list.bal]: Don't like SetOnCopy. Maybe fixedArrayReplace

File: compiler/modules/types/list.bal [hunk 14]
```diff
@@ -377,7 +398,7 @@ function listConjunction(Context cx, Conjunction? con) returns ListConjunction?
     return ();
 }
 
-function bddListMemberType(Context cx, Bdd b, int? key, SemType accum) returns SemType {
+function bddListMemberType(Context cx, Bdd b, IntSubtype|boolean key, SemType accum) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/list.bal]: |true

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/types/list.bal [hunk 15]
```diff
@@ -265,7 +273,12 @@ function listInhabited(Context cx, FixedLengthArray members, SemType rest, ListC
             if listInhabited(cx, members, NEVER, neg.next) {
                 return true;
             }
-            foreach int i in len + 1 ..< negLen {
+            // If last member of neg repeats, just checking one repeating occurrence should suffice.
```
Comment (jclark) [inline compiler/modules/types/list.bal:276]: I don't understand why are not considering neg.maxInitialLen here.

Even if nt is the same after negLenLimit, the rest of the list may vary on each index up to neg.maxInitialLen.

**Review by jclark — CHANGES_REQUESTED** [review]

**Pull request merged**